Greenplum named restore points

jimmyyih opened the pull request:
In Postgres, a named restore point helps with Point-In-Time
Recovery. They are created by calling pg_create_restore_point(), and
are used by setting them in replica recovery.conf files to restore the
database up to that point. This is technically the same in Greenplum
but we have to take care of the distributed part and keep in mind
cluster-wide consistency (e.g. taking restore points during concurrent
twophase commits).

To do this, we introduce a new catalog function
gp_create_restore_point(). It acquires a new lwlock,
TwophaseCommitLock, in exclusive mode to prevent any concurrent
twophase commits from happening in CommitTransaction. When the lock is
acquired, it simply calls pg_create_restore_point() on the QD and QE
segments. After the restore points are created, the lock is released
and any concurrent twophase commits can continue. On the other hand,
QD processes will acquire the new lwlock in shared mode before they
broadcast distributed commit prepared messages to their respective QE
processes.

~~To demonstrate how Greenplum named restore points work, a new test
directory src/test/gpdb_pitr has been added. The test showcases how
the gpdemo HA mirrors can be restored to a certain point before some
DML had occurred.~~

Edit: Changed the test to be actual PITR.
To demonstrate how Greenplum named restore points work, a new test
directory src/test/gpdb_pitr has been added. The test showcases WAL
Archiving in conjunction with the named restore points to do
Point-In-Time Recovery.

yydzero commented:
Great work!

Do we have any estimation on the performance impacts for original cluster?

hlinnaka commented on src/backend/access/transam/xact.c:
```diff
@@ -2805,6 +2815,9 @@ CommitTransaction(void)
 		ProcArrayEndTransaction(MyProc, latestXid, false);
 	}
 
+	if (LWLockHeldByMe(TwophaseCommitLock))
```
The comment in LWLockHeldByMe says `This is meant as debug support only.` It works, but currently it's only used in assertions. It seems unwise to start using it for other things. I'd suggest setting a new local variable to remember if the lock was acquired.

hlinnaka commented on src/backend/access/transam/xact.c:
```diff
@@ -2744,6 +2744,16 @@ CommitTransaction(void)
+	 * GPDB: Acquire TwophaseCommitLock to block a GPDB restore point from
+	 * being created while commit prepared messages are being broadcasted.
+	 */
+	if (s->executorDidWriteXLog && IS_QUERY_DISPATCHER()
```
I think it would be better to do this a bit earlier. The comment at `AtEOXact_RelationMap(true)` says that it should be done as late as possible, so it would seem better to acquire TwophaseCommitLock before that.

hlinnaka commented on src/include/catalog/pg_proc.sql:
```diff
@@ -459,3 +459,5 @@ CREATE FUNCTION cdblegacyhash_cash(money) RETURNS int4 LANGUAGE internal IMMUTAB
 CREATE FUNCTION cdblegacyhash_uuid(uuid) RETURNS int4 LANGUAGE internal IMMUTABLE STRICT AS 'cdblegacyhash_uuid' WITH (OID=6170, DESCRIPTION="Legacy cdbhash function");
 CREATE FUNCTION cdblegacyhash_anyenum(anyenum) RETURNS int4 LANGUAGE internal IMMUTABLE STRICT AS 'cdblegacyhash_anyenum' WITH (OID=6171, DESCRIPTION="Legacy cdbhash function");
+
+CREATE FUNCTION gp_create_restore_point(text) RETURNS void LANGUAGE internal IMMUTABLE STRICAT AS 'gp_create_restore_point' WITH (OID=6998,  DESCRIPTION="Create a named restore point on all segments");
```
Typo: STRICAT

hlinnaka commented on src/backend/access/transam/xlogfuncs.c:
```diff
@@ -529,3 +533,39 @@ pg_backup_start_time(PG_FUNCTION_ARGS)
+		elog(ERROR, "cannot use gp_create_restore_point() when not in QD mode");
+
+	restore_command = psprintf("SELECT pg_catalog.pg_create_restore_point('%s')",
+								   quote_identifier(restore_name_str));
```
Should use `quote_literal_cstr`, not `quote_identifier`.

jimmyyih commented:
@hlinnaka Thanks for the corrections.  As for your scenario, we have `gp_create_restore_point()` releasing the TwophaseCommitLock (exclusive mode) at the end of its execution so it should technically be available for shared mode when the transaction ultimately finishes at `CommitTransaction()` .  Putting some debug logs around the lock acquire/release points, that seems to be correct from log perspective.

jimmyyih commented:
@yydzero We only have some logical guesses for common scenarios.
1. No transactions running.  The restore point records (one record for each segment) will be created with no issues on all segments and should be very fast.
2. Twophase transactions are running.  Depending on who gets the lock first, either the distributed commit will block or the restore point record creation will block.  Either way, it shouldn't have that much impact since distributed commit prepared phase is fast (creates one commit prepared record and updates pg_distributedlog for each segment) and the restore point record creation is also fast (creates one restore point record for each segment).

hlinnaka commented:
>  As for your scenario, we have `gp_create_restore_point()` releasing the TwophaseCommitLock (exclusive mode) at the end of its execution so it should technically be available for shared mode when the transaction ultimately finishes at `CommitTransaction()` . Putting some debug logs around the lock acquire/release points, that seems to be correct from log perspective.

Ah, gotcha, I was confused.

soumyadeep2007 commented on src/backend/access/transam/xlogfuncs.c:
```diff
@@ -529,3 +533,39 @@ pg_backup_start_time(PG_FUNCTION_ARGS)
+/*
+ * gp_create_restore_point: a distributed named point for cluster restore
+ */
+void
```
We should return something. At least a `PG_RETURN_BOOL(true)` perhaps?
We could go one step further and return tuples of the form:
`(gp_segment_id, LSN_OF_RESTORE_POINT)`

jimmyyih replied:
I originally planned to have it return tuples as you have described but converting the function to be an SRF (set return function) was taking too much time and kept getting more and more complex.  I needed to include many CDB front-end headers to this primarily postgres backend file and the dispatch results didn't include the segment id for each individual result (it's in an array so I guess I could have assumed the array index was the content id?).  Since it was getting too complicated, I decided to scrap the logic and keep it as simple as possible.  Also, the LSN returned by `pg_create_restore_point()` is not what you would expect and incorrect IMO...  it's actually the start LSN of the next record (or end LSN of the restore point record).  Doing a grep on xlog dumps with those LSN records wouldn't give you the restore point records.

As for returning a boolean True, I don't think that's necessary.  If it returns without error, that's basically a boolean True.  I don't see where we would return a boolean False instead of just giving an error.  Are you thinking it could be used as some conditional in a SQL query?

soumyadeep2007 replied:
> I originally planned to have it return tuples as you have described but converting the function to be an SRF (set return function) was taking too much time and kept getting more and more complex. I needed to include many CDB front-end headers to this primarily postgres backend file and the dispatch results didn't include the segment id for each individual result (it's in an array so I guess I could have assumed the array index was the content id?). Since it was getting too complicated, I decided to scrap the logic and keep it as simple as possible. Also, the LSN returned by `pg_create_restore_point()` is not what you would expect and incorrect IMO... it's actually the start LSN of the next record (or end LSN of the restore point record). Doing a grep on xlog dumps with those LSN records wouldn't give you the restore point records.
Agreed.
> As for returning a boolean True, I don't think that's necessary. If it returns without error, that's basically a boolean True. I don't see where we would return a boolean False instead of just giving an error. Are you thinking it could be used as some conditional in a SQL query?

I was thinking of providing some feedback to the user, right now it is just an empty response. Usually I see it as a pattern to return 
PG_RETURN_BOOL(true) in many functions...such as gp_request_fts_probe_scan, bminsert, pg_rotate_logfile etc. Yes it would also open up the scope for using it as a conditional in query rather than relying on whether it errors out.

hlinnaka replied:
pg_create_restore_point() returns the (end) LSN of the restore point record. With that, the client knows how much WAL it needs to stream/archive, to get to the restore point. How will that work with gp_create_restore_point()? What will the client logic look like?

jimmyyih replied:
Presumably, client logic would only be dealing with the restore point name given to `gp_create_restore_point()` since it can easily be used with recovery.conf param `recovery_target_name`.  On a large production cluster (say 300+ segments), I would think that storing so many LSN values would be cumbersome vs. just using the restore point name which is consistent across the cluster.  We could have a user-facing utility similar to `gpconfig` which sets the param (and then deprecate the utility once the param moves to postgresql.conf in the later Postgres versions).  Yes, the restore point names are not as unique as LSN values but I think it should be fine to leave it up to the user for now or even develop some automated way later.  Is this an okay approach or is there a better one we should evaluate?

soumyadeep2007 replied:
There should be an API (maybe in a later PR) exposed to the user that lets them see what restore points they currently have created with `gp_create_restore_point()` and are actually "reachable" for a recovery (i.e. what if the xlog was truncated.....).
Also, as for uniqueness of provided restore point names, we can defer to PG behavior/expectations:
"Avoid creating multiple restore points with the same name, since recovery will stop at the first one whose name matches the recovery target."

hlinnaka replied:
> Presumably, client logic would only be dealing with the restore point name given to `gp_create_restore_point()` since it can easily be used with recovery.conf param `recovery_target_name`. On a large production cluster (say 300+ segments), I would think that storing so many LSN values would be cumbersome vs. just using the restore point name which is consistent across the cluster.

The point of returning the LSN is that you can use the LSN to determine how much WAL you need to have archived, so that you can reach the restore point. The scenario in PostgreSQL would be something like this:

--
You have a PostgreSQL server running, and a super-duper WAL archiving setup where the WAL files are shipped to a different data center, and a full base backup is taken each night. The WAL archiving typically lags behind the master by a few seconds, but could be more if there's a network problem, for example. You want to make a restore point, and fire up a new server at that point, using the WAL archive. 

To do that, you first call `pg_create_restore_point('my_restore_point')`, and make note of the LSN it returns. Then, *you wait until the WAL file containing that LSN appears in the WAL archive*. After that, you extract the previous night's base backup to the new server, and create a recovery.conf file in it. In the recovery.conf file, you put a `restore_command` to point to the WAL archive, and `recovery_target_name='my_restore_point`.
--

So the LSN returned by pg_create_restore_point() is not used as the target in recovery.conf. (Although you could do that, too). It's used to ensure that you have archived/streamed  all the WAL required to reach that point. There are other ways to achieve that; instead of using the LSN required by pg_create_restore_point(), you could check that the standby is at least once fully caught-up after creating the restore point, for example. So I guess it's not strictly required to return the LSNs. But what is your plan to achieve that in GPDB?

jimmyyih replied:
> There should be an API (maybe in a later PR) exposed to the user that lets them see what restore points they currently have created with gp_create_restore_point() and are actually "reachable" for a recovery (i.e. what if the xlog was truncated.....).

I guess later we could add a Greenplum catalog table which tracks all the successful distributed restore point creations, and having a "reachable for recovery" true/false state could be doable if we utilize with replication slots.  I wonder how useful that would be in a DR scenario though if the table itself cannot be read due to the main cluster/hosts being offline which is why we went with putting the burden of tracking these restore points to the user.  Either way, I do see value tracking it ourselves.
> 
> Also, as for uniqueness of provided restore point names, we can defer to PG behavior/expectations:
> "Avoid creating multiple restore points with the same name, since recovery will stop at the first one whose name matches the recovery target."

Yup, that was the plan.

jimmyyih replied:
> So the LSN returned by pg_create_restore_point() is not used as the target in recovery.conf. (Although you could do that, too). It's used to ensure that you have archived/streamed all the WAL required to reach that point. There are other ways to achieve that; instead of using the LSN required by pg_create_restore_point(), you could check that the standby is at least once fully caught-up after creating the restore point, for example. So I guess it's not strictly required to return the LSNs. But what is your plan to achieve that in GPDB?

Thanks for the brief explanation!  That was very useful.  The plan for Greenplum is still in research mode (we were going to build ideas on top of this PR)... but some general early ideas we envisioned:
1. Do WAL Archiving and PITR for master and each primary.  Have a separate entity (e.g. PITR coordinator daemon) track restore points by continuously parsing the WAL segment files as they arrive in the archive directory.  It would index which restore points have completely been archived on all segments and mark it as usable.  To execute restore, it'd be similar to Postgres where we create the recovery.conf files and input the `restore_command` and `recovery_target_name` configurations.
2. Similar to (1) but the replicas have been started up and are in continuous recovery.  The coordinator daemon would do the same thing in parsing the WAL segment files as they arrive in the archive directories and indexing restore points to mark for usability.  However, as soon as a distributed restore point is marked as usable, it would move all the WAL segment files (from the previous restore point up to latest restore point) into the replica's pg_xlog directory for the replica to consume.  This would give low downtime as a "HOT" cluster ready for DR failover (and could also be used for a read-only cluster replica).

jimmyyih replied:
The function has been updated to return a boolean but we'll probably make it into an SRF in a later PR to get the LSN values.  I also refactored the test script to utilize basebackups and WAL Archiving to demonstrate PITR (similar to idea 1).

jimmyyih commented on src/test/gpdb_pitr/expected/gpdb_pitr_setup.out:
```diff
@@ -0,0 +1,93 @@
+-- Set a checkpoint and then disable automatic checkpoints
```
Updated the comment to be more clear of the intention.  Essentially, checkpoints when applied will invalidate all previous restore points since everything has been flushed... and each segment individually does automatic checkpoints every 5 minutes.  Needed to disable it to make sure the test is reliable.

But yes, this means that recovery solutions built on top of this PR will have to deal with this (e.g. WAL archiving PITR is relatively safe since WAL is applied on top of a basebackup but a streaming solution would need to handle this).

jimmyyih commented on src/backend/access/transam/xlogfuncs.c:
```diff
@@ -529,3 +533,39 @@ pg_backup_start_time(PG_FUNCTION_ARGS)
+
+	SIMPLE_FAULT_INJECTOR("gp_create_restore_point_acquired_lock");
+
+	CdbDispatchCommand(restore_command, DF_CANCEL_ON_ERROR, NULL);
```
Technically, this shouldn't matter since the dispatch happens before the master writes the restore point record.  If the master doesn't have the restore point, the restore points on the QE segments should be invalid.  However, I only dismissed DF_NEED_TWO_PHASE because I forgot that it could be used in this way to not flush the record on error/abort... so added it to the dispatch flags.  Thanks!

jimmyyih commented on src/backend/access/transam/xact.c:
```diff
@@ -2735,6 +2736,18 @@ CommitTransaction(void)
+	 * point from being created while commit prepared messages are being
+	 * broadcasted.
+	 */
+	if (s->executorDidWriteXLog && IS_QUERY_DISPATCHER()
```
We originally wanted a GUC as well but we thought it to be out of scope for this PR.  But since you mentioned it...

The plan was to guard this pretty invasive conditional with a simple boolean GUC (gp_enable_restore_points as you have suggested) and make it require a cluster restart.  We would also need to do the GUC check in `gp_create_restore_point()` and possibly in `pg_create_restore_point()` if we wanted to be extra careful (but probably not necessary).  It wouldn't be able to be a reloadable GUC since there would be a possibility of a distributed restore point that does not guarantee cluster consistency.  What do you think?

soumyadeep2007 replied:
I would vote for the GUC being a part of this PR if its not too much work to add in. 

The GUC would default to off.

+1 to doing the GUC check in gp_create_restore_point.  

As for doing the GUC check in `pg_create_restore_point`, probably overkill. I don't know if there are cases/arguments for letting people run `pg_create_restore_point` outside the context of a `gp_create_restore_point`. If there are , we would be stepping on the toes of those use cases if we are exiting in `pg_create_restore_point` based on the GUC. We can address this later as well.

As for the GUC level, I was also thinking about a GUC that forces a restart as I was unsure as to whether any other level would work. I don't know how `gpconfig` works, but if we were to use `gpconfig` to set `gp_enable_restore_points `, won't it set it on all the segments with some guarantee of atomicity? How then would we have an inconsistent distributed restore point? What case of inconsistency were you hinting at?

hlinnaka replied:
-1 for adding a GUC. Acquiring a lwlock in shared mode is very cheap. Even though transaction commit is a very hot codepath, I doubt the performance impact is noticeable.

jimmyyih replied:
> As for the GUC level, I was also thinking about a GUC that forces a restart as I was unsure as to whether any other level would work. I don't know how gpconfig works, but if we were to use gpconfig to set gp_enable_restore_points , won't it set it on all the segments with some guarantee of atomicity? How then would we have an inconsistent distributed restore point? What case of inconsistency were you hinting at?

If we made it loadable via `pg_ctl reload`, you could have ongoing concurrent distributed commit broadcasts that do not have the lock acquired and creating a distributed restore point at that moment would not guarantee cluster consistency.

> -1 for adding a GUC. Acquiring a lwlock in shared mode is very cheap. Even though transaction commit is a very hot codepath, I doubt the performance impact is noticeable.

Discussing with @soumyadeep2007 offline, we agreed to not implement a GUC.  If anything, we could always add it later if we do see a significant performance impact... although even if that was the case, it'll probably go into a different direction to see why performance got hit.

hlinnaka commented on src/test/gpdb_pitr/expected/gpdb_pitr_setup.out:
```diff
@@ -0,0 +1,96 @@
+-- Set a checkpoint and then disable any further checkpoints (mainly
+-- targeting automatic checkpoints). Checkpoints can invalidate our
+-- restore points when the minimum recovery ending location in
+-- pg_control file is updated to the checkpoint location.
```
1. I don't understand how a checkpoint would invalidate the restore points. The minimum recovery point does not depend on checkpoints; the minimum recovery point is updated in a standby whenever a buffer is flushed.

2. Even if we need to prevent checkpoints, isn't it enough to run a checkpoint manually? I guess in theory, it's possible that an automatic checkpoint is triggered if checkpoint_timeout is exceeded, but even with a very short timeout like 5 minutes, is it really expected that creating a few tables and inserting a few rows, like this test does, will take longer than 5 minutes?

jimmyyih replied:
> I don't understand how a checkpoint would invalidate the restore points. The minimum recovery point does not depend on checkpoints; the minimum recovery point is updated in a standby whenever a buffer is flushed.

Yes sorry, the way I wrote it is definitely misleading and not 100% correct after I looked into it more.  The scenario I was trying to note down is when an automatic checkpoint record is streamed over to the standby and the standby's checkpointer process does its `CreateRestartPoint()` logic when checkpoint_timeout has exceeded.  Even when the cluster is idle, the minimum recovery point in the standby's control file will continue updating due to the aforementioned flow.  Does that sound more correct?

> Even if we need to prevent checkpoints, isn't it enough to run a checkpoint manually? Do we really need to disable automatic checkpoints with fault injection? I guess in theory, it's possible that an automatic checkpoint is triggered if checkpoint_timeout is exceeded, but even with a very short timeout like 5 minutes, is it really expected that creating a few tables and inserting a few rows, like this test does, will take longer than 5 minutes? Maybe raise checkpoint_timeout for this test, if you're really paranoid about that, we need to change settings before the test, anyway.

That's definitely a good point that I didn't think about since I was so focused on preventing the minimum recovery point from updating.  The explicit CHECKPOINT at the beginning of the setup test should suffice.  Thanks.

jimmyyih replied:
After refactoring the test script to use basebackups and WAL Archiving, the note I had around minimum recovery point is no longer needed.

ashwinstar commented on src/backend/access/transam/xact.c:
```diff
@@ -2735,6 +2736,18 @@ CommitTransaction(void)
+		&& getCurrentDtxState() != DTX_STATE_ONE_PHASE_COMMIT)
+	{
+		LWLockAcquire(TwophaseCommitLock, LW_SHARED);
+		twophaseCommitLockAcquired = true;
```
Do we need to hold this lock before `RecordTransactionCommit()`? I feel we can have this lock after recording the commit on master, just need it before broadcasting commit-prepared to segments. The lock is needed just to not have any commit-prepared and commit-forget records, while creating restore points I think.

With that change, we will not loose this commit, if promoting the cluster post replay to this restore point, as master would re-broadcast the commit to all the segments. With current code if we restore to the restore-point and promote, this transaction will get aborted, which seems avoidable.

Am I missing some case where we need to hold lock across the master's and segments commits?

jimmyyih replied:
That makes sense.  Should I move the lock acquire/release to `doNotifyingCommitPrepared()` instead then?  Like acquire the lwlock before `currentDtxDispatchProtocolCommand()` and release after `doInsertForgetCommitted()`.

jimmyyih replied:
I moved the lwlock acquire/release into `doNotifyingCommitPrepared()` and refactored the tests accordingly to show both a transaction being rebroadcasted due to distributed commit record being written to master and a transaction being automatically aborted with abort prepared when crash recovery is completed (the primary doesn't get a commit prepared signal from master so it assumes the transaction can be aborted).

jimmyyih commented:
@berlin-ab Your comments on the commit seemed to have become removed from this PR after I force pushed a couple times to address the comments from Deep and Heikki.  Luckily, I had your comments saved in a browser tab so I'll record them here with responses.

> This logic seems out of place when surrounded by many of these hook functions. Does it belong in a hook?
 
Yes, the locking logic in `CommitTransaction` does indeed look out of place now that you mention it.  Based on Ashwin's feedback though, I'll most likely be moving the logic more near the distributed transaction broadcasting side.

> an intention explaining function/macro would be nice for the reader here. Of course one phase commit isn't appropriate here, but what is intended - any other dtx state?

Sure, the three conditionals needed to trigger lock acquire doesn't look that clean here but it's more clear and direct rather than passing around stuff into functions/macros (e.g. an executor wrote xlog, the process is a query dispatcher, and it's not a one phase commit).  However, the comment I added above it could use more clarification which I'll add as I look into Ashwin's feedback a bit closer.  If Ashwin is right, we can remove most if not all the conditionals.

> The PR description describes this as being an exclusive lock.

The PR description says the new catalog function `gp_create_restore_point()` acquires the lock in exclusive mode, not at transaction commit.  However, I'll add more to the commit message to clarify that we also take a shared lock here.  This may also change based on what I find out from Ashwin's feedback.

> Why is the local state necessary? Can we instead check later if the lock is held by us?

I originally had it as using `LWLockHeldByMe()` just like you suggested but Heikki suggested changing it to what it is now.  Context here: https://github.com/greenplum-db/gpdb/pull/8983#discussion_r342392708.  After going through the code after seeing Heikki's comment, this is the unfortunate norm (I was thinking of maybe doing something about that in upstream Postgres).

> Seems like an opportunity to create a gp specific file.

I have created a new file and moved the function over.  It's what I wanted to do originally but was 50/50 on it.  Thanks for tipping the scale.

berlin-ab commented:
Thanks for the thorough response.

berlin-ab commented on src/test/gpdb_pitr/sql/gpdb_pitr_setup.sql:
```diff
@@ -0,0 +1,65 @@
+-- Call the restore point creation function. This will merely grab the
+-- TwophaseCommit lwlock in EXCLUSIVE mode until the fault is
+-- released.
+4&: INSERT INTO gpdb_restore_points SELECT 'test_restore_point' FROM gp_create_restore_point('test_restore_point');
```
This is an interesting invocation the function to also be used in a select/insert.

I'm still not used to commands being used in combination with a query.

jimmyyih replied:
Yes, this was to implicitly show that the restore point records are recorded immediately during the transaction and not during transaction commit so the inserted record would not show up later when restoring from the archives.  I've appended to the comment to explicitly call out why we do this somewhat strange query.

jimmyyih commented:
> It seems like our test only executes `gp_create_restore_point` during a happy path scenario. Should we test non-happy path scenarios also?

I'm not sure what non-happy path scenarios would be useful here... especially when the test itself can be considered somewhat fragile where one mess up would bring you right back to the beginning (new cluster needed).  Like for example, testing that PITR would fail if a WAL segment file wasn't archived... but that can easily be assumed.  If you're talking about what would happen if `gp_create_restore_point()` errored out and only some segments wrote the restore point record... yes that would indeed break PITR but that's also somewhat easily assumed.  However, I would say these type of error scenarios would be required testing for a PITR utility built on top of this PR since the utility would have to do error handling and validation checks.  That should be easily doable with a Behave-style framework testing the utility.  In fact, some of us kinda consider the test scripts and stuff to be more of a blueprint for a future PITR utility so look forward to seeing something soon maybe (or maybe not)?

berlin-ab commented:
> > It seems like our test only executes `gp_create_restore_point` during a happy path scenario. Should we test non-happy path scenarios also?
> 
> I'm not sure what non-happy path scenarios would be useful here... especially when the test itself can be considered somewhat fragile where one mess up would bring you right back to the beginning (new cluster needed). Like for example, testing that PITR would fail if a WAL segment file wasn't archived... but that can easily be assumed. If you're talking about what would happen if `gp_create_restore_point()` errored out and only some segments wrote the restore point record... yes that would indeed break PITR but that's also somewhat easily assumed. However, I would say these type of error scenarios would be required testing for a PITR utility built on top of this PR since the utility would have to do error handling and validation checks. That should be easily doable with a Behave-style framework testing the utility. In fact, some of us kinda consider the test scripts and stuff to be more of a blueprint for a future PITR utility so look forward to seeing something soon maybe (or maybe not)?

I guess what I was looking for was:

How do we know that `gp_create_restore_point` was the thing that made it work?

jimmyyih commented:
> I guess what I was looking for was:
> 
> How do we know that gp_create_restore_point was the thing that made it work?

Ah, okay.  If `gp_create_restore_point()` didn't do its thing, the PITR cluster would not have recovered from archive and the replica segments would not have come up.  Since they did come up, we can assume it did some sort of recovery... but did they restore up to the restore point, restore up to the last LSN of the last WAL segment file, and are they still in continuous recovery or have they finished recovery (accepting connections)?  That's where the validation SQL test comes into play where it checks all the tables have the expected rows they should have when the restore points are created (which implicitly means the PITR cluster is running like any other GPDB cluster).  It's a mix of reading the test script (tests PITR and WAL Archiving setup and execution) and the SQL (tests commit blocking, distributed restore points, and PITR conclusion).

But to be a bit snarky, yes technically `pg_create_restore_point()` called in utility-mode on all segments at the same very precise timings could give the same results... but that'd be really hard to automate and is a very unlikely scenario.

jimmyyih commented:
I'm planning on merging this PR tomorrow as there have been no new comments and no explicit -1's in the past two weeks.
